torch/csrc/distributed/autograd/context/context.cpp

```diff
@@ -112,13 +124,28 @@ void DistAutogradContext::addOutstandingRpc(
         if (futErr) {
           // If we have an error, let the local autograd engine know about it.
           std::runtime_error err((*futErr).what());
-          graphTask_->set_exception(err, nullptr);
+          std::unique_lock<std::mutex> lock(lock_);
+          if (graphTask_) {
+            auto future_result =
+                graphTask_->set_exception_without_signal(nullptr);
+            lock.unlock();
```

Comment:
The whole point of breaking `set_exception` in two is to be able to release this lock in the middle?
Why is it not enough to make a local copy in this scope of `graphTask_` to make sure it does not get deleted too early?

Reply:
Making the local copy would work, although its nicer to avoid holding a lock when we mark the future as completed. Since the future, would trigger callbacks and its better not to hold locks there. You can see the discussion on https://github.com/pytorch/pytorch/pull/33156.

Reply:
The other reason for breaking up set_exception is to avoid the mutex. If we don't unlock here, the callbacks called from future->setError will try to lock the mutex again.